utils/signer/mssign.c

```diff
@@ -0,0 +1,183 @@
+#include "common/curl-p/trit.h"
+#include "common/kerl/hash.h"
+#include "common/sign/v2/iss_curl.h"
+#include "common/trinary/trit_tryte.h"
+#include "mam/v1/merkle.h"
+
+#include <getopt.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#define CHARGS "d:s:i:x:t:"
+
+static struct option long_options[] = {
+    {"message", required_argument, NULL, 'm'},
+    {"seed", required_argument, NULL, 's'},
+    {"index", required_argument, NULL, 'i'},
+    {"sec-level", required_argument, NULL, 'x'},
+    {"tree", required_argument, NULL, 't'},
+    {NULL, 0, NULL, 0}};
+
+typedef struct {
+  int key_i;
+  int64_t offset;
+  size_t sec;
+  trit_t *msg;
+  size_t msg_len;
+  trit_t seed[HASH_LENGTH];
+  trit_t *tree;
+  size_t tree_len;
+  size_t n_leafs;
+} ToSign;
+
+int test_hash(trit_t *hash, size_t security) {
+  size_t i;
+  size_t sum = 0;
+
+  for (i = 0; i < HASH_LENGTH / 3; i++) {
+    sum += hash[i];
+  }
+  if (security == 1 && sum != 0) {
+    return 1;
+  }
+  for (i = HASH_LENGTH / 3; i < 2 * HASH_LENGTH / 3; i++) {
+    sum += hash[i];
+  }
+  if (security == 2 && sum != 0) {
+    return 1;
+  }
+  for (i = 2 * HASH_LENGTH / 3; i < HASH_LENGTH; i++) {
+    sum += hash[i];
+  }
+  return sum != 0;
+}
+
+int increment_trits(trit_t *t, size_t length) {
+  trit_t *end = &t[length];
+  trit_t c = 0;
+  for (; t < end; t++) {
+    (*t) += 1 + c;
+    c = *t > 1;
+    if (c) {
+      *t = -1;
+    } else {
+      break;
+    }
+  }
+  return t == end;
+}
+
+int find_nonce(ToSign *args, trit_t *hash) {
+  trit_t nonce[HASH_LENGTH] = {0};
+  tryte_t nonce_str[(HASH_LENGTH / 9) + 1] = {0};
+  Kerl k, k_copy;
+  init_kerl(&k);
+  kerl_absorb(&k, args->msg, args->msg_len);
```

Comment:
This will assert if `args->msg_len % HASH_LENGTH != 0`, do we assume the message has this kind of length ?